spacy_experimental/span_finder/span_finder_component.py

```diff
@@ -0,0 +1,327 @@
+width = ${model.tok2vec.embed.width}
+window_size = 1
+maxout_pieces = 3
+depth = 4
+"""
+
+DEFAULT_SPAN_FINDER_MODEL = Config().from_str(span_finder_default_config)["model"]
+DEFAULT_CANDIDATES_KEY = "span_candidates"
+
+
+@Language.factory(
+    "experimental_span_finder",
+    assigns=["doc.spans"],
+    default_config={
+        "threshold": 0.3,
+        "model": DEFAULT_SPAN_FINDER_MODEL,
+        "candidates_key": DEFAULT_CANDIDATES_KEY,
+        "max_length": 0,
+        "min_length": 0,
+        "scorer": {
+            "@scorers": "spacy-experimental.span_finder_scorer.v1",
+            "candidates_key": DEFAULT_CANDIDATES_KEY,
+        },
+    },
+    default_score_weights={
+        "span_finder_f": 1.0,
+        "span_finder_p": 0.0,
+        "span_finder_r": 0.0,
+    },
+)
+def make_span_finder(
+    nlp: Language,
+    name: str,
+    model: Model[List[Doc], Floats2d],
+    scorer: Optional[Callable],
+    threshold: float,
+    max_length: int,
+    min_length: int,
+    candidates_key: str = DEFAULT_CANDIDATES_KEY,
+) -> "SpanFinder":
+    """Create a SpanFinder component. The component predicts whether a token is the start or the end of a potential span.
+    model (Model[List[Doc], Floats2d]): A model instance that
+        is given a list of documents and predicts a probability for each token.
+    threshold (float): Minimum probability to consider a prediction positive.
+    candidates_key (str): Name of the SpanGroup the predicted spans are saved to
+    max_length (int): Max length of the produced spans (no max limitation when set to 0)
+    min_length (int): Min length of the produced spans (no min limitation when set to 0)
+    """
+    return SpanFinder(
+        nlp.vocab,
+        model=model,
+        threshold=threshold,
+        name=name,
+        scorer=scorer,
+        max_length=max_length,
+        min_length=min_length,
+        candidates_key=candidates_key,
+    )
+
+
+def make_span_finder_scorer(candidates_key: str = DEFAULT_CANDIDATES_KEY):
+    def span_finder_score(examples: Iterable[Example], **kwargs) -> Dict[str, Any]:
+
+        references = get_span_references([doc.reference for doc in examples])
+        predictions = get_span_predictions(
+            [doc.predicted for doc in examples], candidates_key
+        )
+
+        scorer = PRFScore()
+
+        for prediction_doc, reference_doc in zip(predictions, references):
+            for prediction, reference in zip(prediction_doc, reference_doc):
+                if prediction in reference_doc:
+                    scorer.tp += 1
+                else:
+                    scorer.fp += 1
+
+                if reference not in prediction_doc:
+                    scorer.fn += 1
+
+        # Assemble final result
+        final_scores: Dict[str, Any] = {
+            f"span_finder_f": scorer.fscore,
+            f"span_finder_r": scorer.precision,
+            f"span_finder_p": scorer.recall,
+        }
```

Comment:
Hmm, I don't think this needs a custom scorer? This seems like it should be extremely similar to `spancat_score` just unlabeled?

Reply:
After looking a bit more, I understand a bit more why you implemented it this way, but it's not going to work in `get_span_references` or in `_get_reference` to use all the spans in the reference document rather than specifying a particular spans key to learn from. Let me think a bit more about how this could work...